components/ILIAS/Chatroom/classes/CleanUpBans.php

```diff
@@ -0,0 +1,37 @@
+<?php
+
+/**
+ * This file is part of ILIAS, a powerful learning management system
+ * published by ILIAS open source e-Learning e.V.
+ *
+ * ILIAS is licensed with the GPL-3.0,
+ * see https://www.gnu.org/licenses/gpl-3.0.en.html
+ * You should have received a copy of said license along with the
+ * source code, too.
+ *
+ * If this is not the case or you just want to try ILIAS, you'll find
+ * us at:
+ * https://www.ilias.de
+ * https://github.com/ILIAS-eLearning
+ *
+ *********************************************************************/
+
+declare(strict_types=1);
+
+namespace ILIAS\Chatroom;
+
+use ilDBInterface;
+
+class CleanUpBans
+{
+    public function __construct(private readonly ilDBInterface $db)
+    {
+    }
+
+    public function removeEntriesOfDeletedUsers(): void
+    {
+        $this->db->manipulate(
```

Comment:
The statement is executed whenever the `deleteUser` event was fired/raised. Have you done any performance benchmarking for cases where dozens or even hundreds of users are deleted at once?
And: IMO a `JOIN` instead of the `IN()` + `SUBQUERY` performs better.

Options:

- Only delete the record matching the user id of the fired/raised event, together with the usage of a cached (statically) prepared statement. This could be done in a `CachedCleanupBans` file, or even better (but this would be a bit overengineering) by a separate repository used by `CleanupBans`. In this case, the SQL operation should be moved to a `BanRepository` and decorated this with a `CachedBanRepository` (which simply caches the statement statically).
-  We should use a (optimized, see my comment above) version of your proposal by ensuring the statement is executed only once per request.

Reply:
> AFAIK I cannot use join when using delete.

That's not true, `JOIN` can be used in a `DELETE` statement.

In general, there are IMO three approaches when deleting records of a table based on data of a second relation:

- `IN` or `=` with sub query
- `EXISTS` with sub query
- `JOIN`

As you already did some benchmarking, please use the fastest approach here.

Reply:
Hi @mjansenDatabay,
this code path now uses prepared statements with the given user id. So the subquery or join discussion isn't relevant anymore.

It was the fastest approach.